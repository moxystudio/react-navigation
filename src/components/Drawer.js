/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
import React, { useMemo } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import { useNavigation } from '../hooks';

const Drawer = ({ className, navigationItems, LinkComponent, placement }) => {
    const { isDrawerOpen, toggleDrawer } = useNavigation();

    const drawerClassName = useMemo(
        () => classNames(
            'react-navigation__drawer',
            placement,
            { show: isDrawerOpen },
            className,
        ),
        [className, isDrawerOpen, placement],
    );

    return (
        <div data-testid="drawer" className={ drawerClassName }>
            <ul>
                {navigationItems.map(({ href, text }, idx) => (
                    <li key={ idx } onClick={ toggleDrawer }>
                        <LinkComponent href={ href }>
                            <a>{text}</a>
                        </LinkComponent>
                    </li>
                ))}
            </ul>
        </div>
    );
};

Drawer.propTypes = {
    className: PropTypes.string,
    LinkComponent: PropTypes.func.isRequired,
    navigationItems: PropTypes.arrayOf(
        PropTypes.exact({
            href: PropTypes.string,
            text: PropTypes.string,
        }),
    ).isRequired,
    placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
};

Drawer.defaultProps = {
    placement: 'left',
};

export default Drawer;
