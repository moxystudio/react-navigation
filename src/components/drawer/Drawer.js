import React, { useCallback, useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

import { useNavigation } from '../../hooks';

const drawerClassName = 'react-navigation_drawer';

const Drawer = ({
    placement,
    withOverlay,
    lockBodyScroll,
    children,
    className,
    overlayClassName,
}) => {
    const drawerRef = useRef();
    const { drawer } = useNavigation();

    const drawerClassNames = classNames(drawerClassName, `${drawerClassName}-${placement}`, className);
    const overlayClassNames = classNames(`${drawerClassName}-overlay`, overlayClassName);

    useEffect(() => {
        if (!lockBodyScroll) { return; }

        if (drawer.isOpen) {
            disableBodyScroll(drawerRef.current);
        } else {
            enableBodyScroll(drawerRef.current);
        }
    }, [drawer.isOpen, lockBodyScroll]);

    useEffect(() => {
        const drawer = drawerRef.current;

        return () => {
            enableBodyScroll(drawer);
        };
    }, []);

    const handleOverlayClick = useCallback(() => {
        drawer.close();
    }, [drawer]);

    return (
        <>
            { withOverlay &&
                <div
                    className={ overlayClassNames }
                    data-open={ drawer.isOpen }
                    data-placement={ placement }
                    onClick={ handleOverlayClick } />
            }
            <div
                ref={ drawerRef }
                className={ drawerClassNames }
                data-open={ drawer.isOpen }
                data-placement={ placement }>
                { children }
            </div>
        </>
    );
};

Drawer.propTypes = {
    placement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
    withOverlay: PropTypes.bool,
    lockBodyScroll: PropTypes.bool,
    children: PropTypes.any,
    className: PropTypes.string,
    overlayClassName: PropTypes.string,
};

Drawer.defaultProps = {
    placement: 'left',
    withOverlay: true,
    lockBodyScroll: true,
};

export default Drawer;
