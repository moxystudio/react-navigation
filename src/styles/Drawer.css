@import "./variables.css";
@import "./custom-medias.css";

.react-navigation_drawer {
    position: absolute;
    width: 80vw;
    height: 100vh;
    z-index: calc(var(--z-index-base) + 2);
    transform: translateX(-100%);
    background-color: var(--color-beige);
    transition: transform var(--duration-standard);

    &.show {
        transform: translateX(0);
    }
}

@media (--layout-lte-xsmall) {
    .react-navigation_drawer {
        width: 100vw;
        z-index: var(--z-index-base);
        transform: translateY(calc(-100% + 5rem));

        &.show {
            /* header height */
            transform: translateY(5rem);
        }
    }
}
