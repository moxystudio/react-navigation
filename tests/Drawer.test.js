import React from 'react';
import { render, screen } from '@testing-library/react';

import Drawer from '../src/components/Drawer';
import AppTree from './AppTree';

const defaultProps = {
    placement: 'left',
    navigationItems: [
        { href: '/lorem', text: 'Lorem' },
        { href: '/ipsum', text: 'Ipsum' },
    ],
    LinkComponent: jest.fn(({ children, href }) => (
        <a href={ href }>
            {children.props.children}
        </a>
    )),
};

const renderWithProps = (props = {}, Component = Drawer) => render(
    <AppTree>
        <Component { ...defaultProps } { ...props } />
    </AppTree>,
);

describe('Drawer', () => {
    it('should render correctly', () => {
        const { asFragment } = renderWithProps();

        expect(asFragment()).toMatchSnapshot();
    });

    it('should render correctly with className', () => {
        renderWithProps({ className: 'foo' });

        expect(screen.getByTestId('drawer')).toHaveClass('foo');
    });

    it('should be invisible by default', () => {
        renderWithProps();

        expect(screen.getByTestId('drawer')).not.toHaveClass('show');
    });

    it('should be visible when the its state is toggled', () => {
        const mockNavigationContextValue = {
            isDrawerOpen: true,
            toggleDrawer: jest.fn(),
        };

        render(
            <AppTree navigationContextValue={ mockNavigationContextValue }>
                <Drawer { ...defaultProps } />
            </AppTree>,
        );

        expect(screen.getByTestId('drawer')).toHaveClass('show');
    });

    it('should render correctly with top placement', () => {
        renderWithProps({ placement: 'top' });

        expect(screen.getByTestId('drawer')).toHaveClass('top');
    });

    it('should render correctly with right placement', () => {
        renderWithProps({ placement: 'right' });

        expect(screen.getByTestId('drawer')).toHaveClass('right');
    });

    it('should render correctly with bottom placement', () => {
        renderWithProps({ placement: 'bottom' });

        expect(screen.getByTestId('drawer')).toHaveClass('bottom');
    });

    it('should render the navigation items', () => {
        renderWithProps();

        const { getByText } = screen;

        expect(getByText(/lorem/i)).toBeInTheDocument();
        expect(getByText(/ipsum/i)).toBeInTheDocument();
    });
});

