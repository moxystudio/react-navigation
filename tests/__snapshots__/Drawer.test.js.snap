// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Drawer should render correctly 1`] = `
<DocumentFragment>
  <div
    class="react-navigation__drawer left"
    data-testid="drawer"
  >
    <ul>
      <li>
        <a
          href="/lorem"
        >
          Lorem
        </a>
      </li>
      <li>
        <a
          href="/ipsum"
        >
          Ipsum
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
